### ISSUE TITLE
Vault loses auth backend tuning settings?

### ISSUE THREAD

**Issue by raffraffraff** [hashicorp/vault#1752]
We have run a pair of Vault servers in each of our datacenters. They're all backed by a local Consul cluster, which they use for HA and storage. We have had several major issues with mass token expiry. When we investigate, it appears that the token backend settings have changed: specifically, 'max_lease_ttl' reverts to 30 days (which I believe is the default).

There doesn't seem to be a pattern. We're currently running Vault 0.5.2 in production, and are currently starting to deploy Vault 0.6.0 (haven't seen the issue there yet, but it's too early to say). I have created a sensu check that alerts me if the values on this API path are not what we expect:

```
v1/sys/mounts/auth/token/tune
```

I just got another alert:  'max_lease_ttl' and 'default_lease_ttl' are 30 days. All tokens that were already older than 30 days were expired. This means that I have to create tokens and rerun the vault auth tune again:

```
vault mount-tune -default-lease-ttl=$DEFAULT -max-lease-ttl=$MAX auth/token
```

- Anyone else experiencing this issue?
- Any tips for troubleshooting this?

**Comment by jefferai** [comment]
Tuning auth backends wasn't supported in 0.5.2 but it is in 0.6.0!

**Comment by raffraffraff** [comment]
Really? That's odd - I can run the vault mount-tune command on my 0.5.2 vault and it works. It retains the setting for a /dev/random amount of time, and then suddenly loses it. 

So the upgrade will fix this? Because that would _rock_

**Comment by jefferai** [comment]
@raffraffraff It doesn't actually work, as you saw. What happens is that the tuning code in 0.5.2 was written for secret backends, but because of the way it works, if you tune an auth backend in 0.5.2 it will behave *on the current active node* like the tuning worked...but the info is not actually persisted, so at the next failover it'll be gone. This is unfortuante UX because it looks like something succeeded when it didn't, not because of a bug, because it wasn't actually implemented.

In 0.6 we explicitly added support for tuning auth backends, meaning, if you attempt to tune an auth backend, this is detected and the tuning information is persisted to the auth mount table. So you should be good!

**Comment by raffraffraff** [comment]
Right, makes sense, and it explains what we see. 

Unfortunately, this makes this our current v0.5.2 production Vault setup very brittle. I'm writing up an RCA on our last failure, to preserve confidence in the solution.  So I should be able to find failover events that coincides with the token expiry, or at least demonstrate that the problem occurs after a vault step-down.

I'm going to accelerate our 0.6.0 deployment, and do some testing on our non-production datacenter to demonstrate that the TTL settings are persistent across failovers.

**Comment by jefferai** [comment]
Hah, sure. Sorry you had trouble with outages.

BTW, 0.6.1 comes out today or Monday, so if you're looking to upgrade, you may want to give it a look. It's a major feature release...tons of improvements and lots of bug fixes.

**Comment by raffraffraff** [comment]
Dang. I'll take a look. Honestly though, I can't leave a ticking time-bomb in our infrastructure over the weekend. I might schedule another upgrade next sprint.

BTW: The reason I assumed that 0.5.2 could cope with auth mount tuning was that it was mentioned in this issue: https://github.com/hashicorp/vault/issues/1079

^ That, and the fact that the vault mount tune lets you do it.

**Comment by jefferai** [comment]
> That, and the fact that the vault mount tune lets you do it.

Yep, that's that bad UX I was talking about. It should have caused an error in 0.5.2 but didn't.

Have you looked at periodic tokens? Those are pretty good choices for servers because they have no max TTL. There are restrictions on how they can be created but the idea is that the TTL is relatively low (say, a few hours) but constantly renewed. So long as the service using the token is alive, it can keep using it; otherwise (if not revoked manually during shutdown) it expires pretty quickly.

**Comment by raffraffraff** [comment]
Yeah, they may suit some of our servers, but others only talk to vault during container restarts, which could be every day, or once in a few months. For now, the focus is to get all secrets into Vault using a simple one-size-fits-all approach, but I'm keeping my eye on the release changelogs.

**Comment by jefferai** [comment]
:+1:

**Comment by narayan8291** [comment]
@raffraffraff : I faced this issue too. I started creating tokens and suddenly the TTL was 30 days and I remember tuning the auth/token backend for a default TTL of 60 days. So, I had to to execute the auth/token tune command whenever a vault in brought back to unsealed state from sealed state. 

@jefferai : Thanks for fixing this bug in 0.6. I will wait for the upgrade. Would 0.6.1 include IP based access ?
